models/mobilenet/mobilenet_v1_impl.hpp

```diff
@@ -0,0 +1,191 @@
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+MobileNetV1<OutputLayerType, InitializationRuleType>::MobileNetV1(
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const float alpha,
+    const size_t depthMultiplier,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    alpha(alpha),
+    depthMultiplier(depthMultiplier),
+    numClasses(numClasses)
+{
+  if (inputWidth < 32 || inputHeight < 32)
+  {
+    mlpack::Log::Fatal << "input width and input height cannot be smaller than"
+        " 32.\nGiven input width and height: (" << inputWidth << ", "
+        << inputHeight << ")" << std::endl;
+  }
+
+  if (preTrained)
+  {
+    if (numClasses != 1000)
+    {
+      mlpack::Log::Fatal << "Number of classes should be 1000 when pre-trained"
+          " is true." << std::endl;
+    }
+
+    if (inputWidth != inputHeight)
+    {
+      mlpack::Log::Fatal << "When pre-trained is true image height should be"
+          " equal to image width." << std::endl;
+    }
+
+    std::map<size_t, std::string>::iterator imageSizeString =
+        imageSizeToString.find(inputWidth);
+    if (imageSizeString == imageSizeToString.end())
+    {
+      mlpack::Log::Fatal << "Image size can only be one of the following when"
+          " pre-trained is true: (128, 160, 192, 224)" << std::endl;
+    }
+
+    std::map<double, std::string>::iterator alphaString =
+        alphaToString.find(alpha);
+    if (alphaString == alphaToString.end())
+    {
+      mlpack::Log::Fatal << "Alpha can only be one of the following when"
+          " pre-trained is true: (0.25, 0.5, 0.75, 1.0)" << std::endl;
+    }
+
+    std::string home = getenv("HOME");
+    preTrainedPath = home + "/.cache/mlpack/models/weights/mobilenetv1/"
+        "mobilenetv1_" + alphaString->second + "_" + imageSizeString->second +
+        ".bin";
+    if (Utils::PathExists(preTrainedPath, true) == false)
+    {
+      std::cout << "mobilenetv1_" + alphaString->second + "_" +
+          imageSizeString->second + ".bin to " + preTrainedPath << std::endl;
+      Utils::DownloadFile("mobilenetv1_" + alphaString->second + "_" +
+          imageSizeString->second + ".bin", preTrainedPath, "", false, false,
+          "http://models.mlpack.org/mobilenetv1/");
+    }
+
+    LoadModel(preTrainedPath);
+    return;
+  }
+  outSize = size_t(32 * alpha);
+  mobileNet.Add(new ann::Convolution<>(inputChannel, outSize, 3, 3, 2, 2,
+      std::make_tuple(0, 1), std::make_tuple(0, 1), inputWidth, inputHeight));
+  mlpack::Log::Info << "Convolution: " << "(" << "3, " << inputWidth + 1 << ", "
+      << inputHeight + 1 << ")" << " ---> (" << outSize << ", ";
+  inputWidth = ConvOutSize(inputWidth, 3, 2, 1);
+  inputHeight = ConvOutSize(inputHeight, 3, 2, 1);
```

Comment:
Lets have a function for this, should be cleaner.

Reply:
Done, i've replaced this and also the conv block inside the deptwise conv block.